torchdata/stateful_dataloader/ibm_rescalable.py

```diff
@@ -0,0 +1,614 @@
+    ----
+    dataset : _StatefulDataset
+        Fully instantiated dataset. Cloned into logical workers during setup fn.
+    n_logical_shards : int
+        Total number of logical shards. Must be a multiple of world size.
+    verbose : bool
+        Track setup progress?
+    """
+
+    def __init__(
+        self,
+        dataset: _StatefulDataset,
+        n_logical_shards: int = 2048,
+        verbose=False,
+    ):
+        super().__init__(dataset)
+        assert (
+            n_logical_shards % self.worldsize == 0
+        ), f"World size {self.worldsize} must divide n_logical_shards {n_logical_shards} evenly"
+        assert n_logical_shards > 0, f"n_logical_shards {n_logical_shards} must be a positive integer"
+
+        self.total_shards = n_logical_shards
+        self.verbose = verbose
+
+        # Fields to be populated during setup / subdataset setup
+        self.data: List[_StatefulDataset] = []
+        self.logicals_owned: List[int] = []
+        self.n_logicals = 0
+
+        # Position "state", used only for maintaining order when n_workers is unchanged
+        # For scaling up or down, logical position is meaningless, and reset
+        self.current_reader = 0
+        self.load_worldsize = self.worldsize
+
+        self.state_params = ["current_reader"]  # self.data states are handled manually
+
+    def setup(self):
+        if not self.is_setup:
+            _StatefulDataset.setup(self)
+            n_logical_shards = self.total_shards
+            logicals = list(range(n_logical_shards))
+            self.logicals_owned = _shard_partition(logicals, self.rank, self.worldsize)
+            self.n_logicals = n_logical_shards // self.worldsize
+            assert (
+                len(self.logicals_owned) == self.n_logicals
+            ), "(world size * num workers) does not divide logical shards evenly"
+
+            # Build logical shards
+            for i in range(self.n_logicals):
+                self.data.append(deepcopy(self.dataset))
+                self.data[-1].worldsize = n_logical_shards
+                self.data[-1].rank = self.logicals_owned[i]
+                self.data[-1].local_worldsize = 1
+                self.data[-1].datapath = self.datapath
+                self.data[-1].verbose = self.rank == 0
+                if self.verbose:
+                    logging.info(
+                        f"Worker {self.rank} assembled logical shard {self.logicals_owned[i]}, {i+1} of {self.n_logicals}"
+                    )
+            [d.setup() for d in self.data]
+
+    def __iter__(self):
+        self.setup()
+        # Grab one item at a time, iterating over owned logical shards
+        data = [iter(d) for d in self.data]
+        while True:
+            ind = self.current_reader
+            # Read doc
+            out = next(data[ind])
+            # Update state
+            self.current_reader = (self.current_reader + 1) % self.n_logicals
+            yield out
+
+    def state_dict(self):
```

Comment:
```
{
  my_children: [c.state_dict() for c in self.children],
  scalar_state: self.scalar, # "my_string"
  my_reshardale_state: tensor.array([1, 2, 3, 4, 5]), # 2d tensor
}

question: what happens if above state_dict gets passed to DCP? 
Answer: it will fail because torch.tensor gets called on everything?

Andrew to follow up with @pradeepfn on this
```